#### Professional data scientists usually spend a very large portion of their time on Data Cleaning. Pandas is the definitive library for performing data analysis with Python. In this notebook we will discuss some tips to do data cleaning in a quick and efficient manner

#### We will see the detailed usage of 

1. DataFrame.loc
2. DataFrame.value_counts
3. Dataframe .query and .memory_usage

## Setup - Import libraries

In [3]:
import pandas as pd
import numpy as np

## Load data and store in dataframe df:

In [4]:
path='https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/automobileEDA.csv'
df = pd.read_csv(path)
df.head()


symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   
3               10.0       102.0    5500.0       24          30  13950.0   
4                8.0       115.0    5500.0       18          22  17450.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  
3     9.791667             Medium       0    1  
4    13.055556             Medium       0    1  

[5 rows x 29 columns]

## 1. Select subset rows and columns using .loc
By using the loc operator, we are able to select subsets of rows and columns on the basis of their index label and column name. Below are some examples on how to use the loc operator on the ‘countries’ DataFrame:

##### Choose all rows  data from multiple consecutive columns indicating 'from' and 'to' column names

In [7]:
df.loc[:, 'symboling':'body-style']

symboling  normalized-losses         make aspiration num-of-doors  \
0            3                122  alfa-romero        std          two   
1            3                122  alfa-romero        std          two   
2            1                122  alfa-romero        std          two   
3            2                164         audi        std         four   
4            2                164         audi        std         four   
..         ...                ...          ...        ...          ...   
196         -1                 95        volvo        std         four   
197         -1                 95        volvo      turbo         four   
198         -1                 95        volvo        std         four   
199         -1                 95        volvo      turbo         four   
200         -1                 95        volvo      turbo         four   

      body-style  
0    convertible  
1    convertible  
2      hatchback  
3          sedan  
4          sedan  
..           ...  
196        sedan  
197        sedan  
198        sedan  
199        sedan  
200        sedan  

[201 rows x 6 columns]

#### We are able to select subsets of rows and columns on the basis of their index label and column name

In [11]:
df.loc[2:10, 'symboling':'body-style']

symboling  normalized-losses         make aspiration num-of-doors  \
2           1                122  alfa-romero        std          two   
3           2                164         audi        std         four   
4           2                164         audi        std         four   
5           2                122         audi        std          two   
6           1                158         audi        std         four   
7           1                122         audi        std         four   
8           1                158         audi      turbo         four   
9           2                192          bmw        std          two   
10          0                192          bmw        std         four   

   body-style  
2   hatchback  
3       sedan  
4       sedan  
5       sedan  
6       sedan  
7       wagon  
8       sedan  
9       sedan  
10      sedan

#### We are able to select particular rows and columns on the basis of their index label and column name

In [13]:
df.loc[[2,10,12], 'symboling':'body-style']


symboling  normalized-losses         make aspiration num-of-doors  \
2           1                122  alfa-romero        std          two   
10          0                192          bmw        std         four   
12          0                188          bmw        std         four   

   body-style  
2   hatchback  
10      sedan  
12      sedan

##### Choose data from multiple columns indicating column names

In [10]:
df.loc[:, ['bore','stroke','compression-ratio','horsepower']]

bore  stroke  compression-ratio  horsepower
0    3.47    2.68                9.0       111.0
1    3.47    2.68                9.0       111.0
2    2.68    3.47                9.0       154.0
3    3.19    3.40               10.0       102.0
4    3.19    3.40                8.0       115.0
..    ...     ...                ...         ...
196  3.78    3.15                9.5       114.0
197  3.78    3.15                8.7       160.0
198  3.58    2.87                8.8       134.0
199  3.01    3.40               23.0       106.0
200  3.78    3.15                9.5       114.0

[201 rows x 4 columns]

## 2. Inclusion - Conditional filtering of DataFrames by category using .loc

For a single category, we are able to do this by using the == operator.

In [14]:
df[df.make == 'alfa-romero']

symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  

[3 rows x 29 columns]

For multiple categories, we have to make use of the isin function:

In [15]:
df[df.make.isin(['alfa-romero', 'audi'])]

symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   
5          2                122         audi        std          two   
6          1                158         audi        std         four   
7          1                122         audi        std         four   
8          1                158         audi      turbo         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   
5        sedan          fwd           front        99.8  0.851994  ...   
6        sedan          fwd           front       105.8  0.925997  ...   
7        wagon          fwd           front       105.8  0.925997  ...   
8        sedan          fwd           front       105.8  0.925997  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   
3               10.0       102.0    5500.0       24          30  13950.0   
4                8.0       115.0    5500.0       18          22  17450.0   
5                8.5       110.0    5500.0       19          25  15250.0   
6                8.5       110.0    5500.0       19          25  17710.0   
7                8.5       110.0    5500.0       19          25  18920.0   
8                8.3       140.0    5500.0       17          20  23875.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  
3     9.791667             Medium       0    1  
4    13.055556             Medium       0    1  
5    12.368421             Medium       0    1  
6    12.368421             Medium       0    1  
7    12.368421             Medium       0    1  
8    13.823529             Medium       0    1  

[9 rows x 29 columns]

## 3. Exclusion - Conditional filtering of DataFrames by category using .loc
Use of the ~ (tilde) sign, which is the complement operator. 

In [16]:
df[~df.make.isin(['alfa-romero', 'audi'])]

symboling  normalized-losses   make aspiration num-of-doors body-style  \
9            2                192    bmw        std          two      sedan   
10           0                192    bmw        std         four      sedan   
11           0                188    bmw        std          two      sedan   
12           0                188    bmw        std         four      sedan   
13           1                122    bmw        std         four      sedan   
..         ...                ...    ...        ...          ...        ...   
196         -1                 95  volvo        std         four      sedan   
197         -1                 95  volvo      turbo         four      sedan   
198         -1                 95  volvo        std         four      sedan   
199         -1                 95  volvo      turbo         four      sedan   
200         -1                 95  volvo      turbo         four      sedan   

    drive-wheels engine-location  wheel-base    length  ...  \
9            rwd           front       101.2  0.849592  ...   
10           rwd           front       101.2  0.849592  ...   
11           rwd           front       101.2  0.849592  ...   
12           rwd           front       101.2  0.849592  ...   
13           rwd           front       103.5  0.908217  ...   
..           ...             ...         ...       ...  ...   
196          rwd           front       109.1  0.907256  ...   
197          rwd           front       109.1  0.907256  ...   
198          rwd           front       109.1  0.907256  ...   
199          rwd           front       109.1  0.907256  ...   
200          rwd           front       109.1  0.907256  ...   

     compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
9                  8.8       101.0    5800.0       23          29  16430.0   
10                 8.8       101.0    5800.0       23          29  16925.0   
11                 9.0       121.0    4250.0       21          28  20970.0   
12                 9.0       121.0    4250.0       21          28  21105.0   
13                 9.0       121.0    4250.0       20          25  24565.0   
..                 ...         ...       ...      ...         ...      ...   
196                9.5       114.0    5400.0       23          28  16845.0   
197                8.7       160.0    5300.0       19          25  19045.0   
198                8.8       134.0    5500.0       18          23  21485.0   
199               23.0       106.0    4800.0       26          27  22470.0   
200                9.5       114.0    5400.0       19          25  22625.0   

    city-L/100km  horsepower-binned  diesel  gas  
9      10.217391                Low       0    1  
10     10.217391                Low       0    1  
11     11.190476             Medium       0    1  
12     11.190476             Medium       0    1  
13     11.750000             Medium       0    1  
..           ...                ...     ...  ...  
196    10.217391             Medium       0    1  
197    12.368421               High       0    1  
198    13.055556             Medium       0    1  
199     9.038462             Medium       1    0  
200    12.368421             Medium       0    1  

[192 rows x 29 columns]

## 4. Reverse row order of data using .loc

In [17]:
df.loc[::-1]

symboling  normalized-losses         make aspiration num-of-doors  \
200         -1                 95        volvo      turbo         four   
199         -1                 95        volvo      turbo         four   
198         -1                 95        volvo        std         four   
197         -1                 95        volvo      turbo         four   
196         -1                 95        volvo        std         four   
..         ...                ...          ...        ...          ...   
4            2                164         audi        std         four   
3            2                164         audi        std         four   
2            1                122  alfa-romero        std          two   
1            3                122  alfa-romero        std          two   
0            3                122  alfa-romero        std          two   

      body-style drive-wheels engine-location  wheel-base    length  ...  \
200        sedan          rwd           front       109.1  0.907256  ...   
199        sedan          rwd           front       109.1  0.907256  ...   
198        sedan          rwd           front       109.1  0.907256  ...   
197        sedan          rwd           front       109.1  0.907256  ...   
196        sedan          rwd           front       109.1  0.907256  ...   
..           ...          ...             ...         ...       ...  ...   
4          sedan          4wd           front        99.4  0.848630  ...   
3          sedan          fwd           front        99.8  0.848630  ...   
2      hatchback          rwd           front        94.5  0.822681  ...   
1    convertible          rwd           front        88.6  0.811148  ...   
0    convertible          rwd           front        88.6  0.811148  ...   

     compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
200                9.5       114.0    5400.0       19          25  22625.0   
199               23.0       106.0    4800.0       26          27  22470.0   
198                8.8       134.0    5500.0       18          23  21485.0   
197                8.7       160.0    5300.0       19          25  19045.0   
196                9.5       114.0    5400.0       23          28  16845.0   
..                 ...         ...       ...      ...         ...      ...   
4                  8.0       115.0    5500.0       18          22  17450.0   
3                 10.0       102.0    5500.0       24          30  13950.0   
2                  9.0       154.0    5000.0       19          26  16500.0   
1                  9.0       111.0    5000.0       21          27  16500.0   
0                  9.0       111.0    5000.0       21          27  13495.0   

    city-L/100km  horsepower-binned  diesel  gas  
200    12.368421             Medium       0    1  
199     9.038462             Medium       1    0  
198    13.055556             Medium       0    1  
197    12.368421               High       0    1  
196    10.217391             Medium       0    1  
..           ...                ...     ...  ...  
4      13.055556             Medium       0    1  
3       9.791667             Medium       0    1  
2      12.368421             Medium       0    1  
1      11.190476             Medium       0    1  
0      11.190476             Medium       0    1  

[201 rows x 29 columns]

## 5. Reverse column order of data using .loc

In [18]:
df.loc[:, ::-1]

gas  diesel horsepower-binned  city-L/100km    price  highway-mpg  \
0      1       0            Medium     11.190476  13495.0           27   
1      1       0            Medium     11.190476  16500.0           27   
2      1       0            Medium     12.368421  16500.0           26   
3      1       0            Medium      9.791667  13950.0           30   
4      1       0            Medium     13.055556  17450.0           22   
..   ...     ...               ...           ...      ...          ...   
196    1       0            Medium     10.217391  16845.0           28   
197    1       0              High     12.368421  19045.0           25   
198    1       0            Medium     13.055556  21485.0           23   
199    0       1            Medium      9.038462  22470.0           27   
200    1       0            Medium     12.368421  22625.0           25   

     city-mpg  peak-rpm  horsepower  compression-ratio  ...    length  \
0          21    5000.0       111.0                9.0  ...  0.811148   
1          21    5000.0       111.0                9.0  ...  0.811148   
2          19    5000.0       154.0                9.0  ...  0.822681   
3          24    5500.0       102.0               10.0  ...  0.848630   
4          18    5500.0       115.0                8.0  ...  0.848630   
..        ...       ...         ...                ...  ...       ...   
196        23    5400.0       114.0                9.5  ...  0.907256   
197        19    5300.0       160.0                8.7  ...  0.907256   
198        18    5500.0       134.0                8.8  ...  0.907256   
199        26    4800.0       106.0               23.0  ...  0.907256   
200        19    5400.0       114.0                9.5  ...  0.907256   

     wheel-base engine-location  drive-wheels   body-style num-of-doors  \
0          88.6           front           rwd  convertible          two   
1          88.6           front           rwd  convertible          two   
2          94.5           front           rwd    hatchback          two   
3          99.8           front           fwd        sedan         four   
4          99.4           front           4wd        sedan         four   
..          ...             ...           ...          ...          ...   
196       109.1           front           rwd        sedan         four   
197       109.1           front           rwd        sedan         four   
198       109.1           front           rwd        sedan         four   
199       109.1           front           rwd        sedan         four   
200       109.1           front           rwd        sedan         four   

     aspiration         make  normalized-losses  symboling  
0           std  alfa-romero                122          3  
1           std  alfa-romero                122          3  
2           std  alfa-romero                122          1  
3           std         audi                164          2  
4           std         audi                164          2  
..          ...          ...                ...        ...  
196         std        volvo                 95         -1  
197       turbo        volvo                 95         -1  
198         std        volvo                 95         -1  
199       turbo        volvo                 95         -1  
200       turbo        volvo                 95         -1  

[201 rows x 29 columns]

## 6. value_counts() Default usage
Returns a Series in descending order of the most frequently-occurring values. By default, rows that contain any NA values are omitted from the result. By default, the resulting Series will be in descending order so that the first element is the most frequently-occurring row.




In [25]:
df['body-style'].value_counts()

sedan          94
hatchback      68
wagon          25
hardtop         8
convertible     6
Name: body-style, dtype: int64

#### We can see how many occurences of each car "body style" we have in the dataset.

#### We can also convert the series to a Dataframe as follows :

In [26]:
df['body-style'].value_counts().to_frame()

body-style
sedan                94
hatchback            68
wagon                25
hardtop               8
convertible           6

## 7. value_counts() in ascending order using ascending parameter
The series returned by value_counts() is in descending order by default. We can reverse the case by setting the ascending parameter to True .

In [30]:
df['body-style'].value_counts(ascending=True)

convertible     6
hardtop         8
wagon          25
hatchback      68
sedan          94
Name: body-style, dtype: int64

## 8. value_counts() using normalize parameter
Return proportions rather than frequencies or number of occurences 

In [29]:
df['body-style'].value_counts(normalize=True)

sedan          0.467662
hatchback      0.338308
wagon          0.124378
hardtop        0.039801
convertible    0.029851
Name: body-style, dtype: float64

#### Of all the information for cars we have in our data set, 46% information is for sedan style cars. We can easily see we have only 3.8% data available for hardtop style and 2.9% data available for convertible style of cars

## 9. value_counts() using bin parameter
for binning of continuous numerical variables into discrete intervals 

In [32]:
# apply value_counts on a numerical column without the bin parameter
df['price'].value_counts()

8495.0     2
18150.0    2
7295.0     2
6229.0     2
8845.0     2
          ..
15580.0    1
6377.0     1
30760.0    1
16925.0    1
18920.0    1
Name: price, Length: 186, dtype: int64

#### This doesnt make much sence as there are too many categories for a particular price of car

In [34]:
df['price'].value_counts(bins=10)

(5077.717, 9146.2]    81
(9146.2, 13174.4]     45
(13174.4, 17202.6]    35
(17202.6, 21230.8]    17
(33315.4, 37343.6]     7
(21230.8, 25259.0]     6
(29287.2, 33315.4]     4
(25259.0, 29287.2]     3
(37343.6, 41371.8]     2
(41371.8, 45400.0]     1
Name: price, dtype: int64

#### We can see price for maximum number of cars (81) are in price range USD 5077 to USD 9146 

## 10. query on specfic columns and data using .query parameter
For those of us who mostly use SQL in our daily work, this function allows for quick and easy access to the subsections of DataFrames that you want to access. 

In [10]:
df.query("price > 35000")

symboling  normalized-losses           make aspiration num-of-doors  \
15           0                122            bmw        std          two   
16           0                122            bmw        std         four   
45           0                122         jaguar        std         four   
46           0                122         jaguar        std          two   
69           3                142  mercedes-benz        std          two   
70           0                122  mercedes-benz        std         four   
71           1                122  mercedes-benz        std          two   
125          3                122        porsche        std          two   

      body-style drive-wheels engine-location  wheel-base    length  ...  \
15         sedan          rwd           front       103.5  0.931283  ...   
16         sedan          rwd           front       110.0  0.946660  ...   
45         sedan          rwd           front       113.0  0.959154  ...   
46         sedan          rwd           front       102.0  0.921192  ...   
69   convertible          rwd           front        96.6  0.866410  ...   
70         sedan          rwd           front       120.9  1.000000  ...   
71       hardtop          rwd           front       112.0  0.957232  ...   
125  convertible          rwd            rear        89.5  0.811629  ...   

     compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
15                 8.0       182.0    5400.0       16          22  41315.0   
16                 8.0       182.0    5400.0       15          20  36880.0   
45                 8.1       176.0    4750.0       15          19  35550.0   
46                11.5       262.0    5000.0       13          17  36000.0   
69                 8.3       155.0    4750.0       16          18  35056.0   
70                 8.0       184.0    4500.0       14          16  40960.0   
71                 8.0       184.0    4500.0       14          16  45400.0   
125                9.5       207.0    5900.0       17          25  37028.0   

    city-L/100km  horsepower-binned  diesel  gas  
15     14.687500               High       0    1  
16     15.666667               High       0    1  
45     15.666667               High       0    1  
46     18.076923                NaN       0    1  
69     14.687500             Medium       0    1  
70     16.785714               High       0    1  
71     16.785714               High       0    1  
125    13.823529               High       0    1  

[8 rows x 29 columns]

In [11]:
query_car_prices = [34184,16430,11694]
df.query("price in @query_car_prices")

symboling  normalized-losses           make aspiration num-of-doors  \
9            2                192            bmw        std          two   
68          -1                122  mercedes-benz        std         four   
145          0                 85         subaru      turbo         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
9        sedan          rwd           front       101.2  0.849592  ...   
68       sedan          rwd           front       115.6  0.973570  ...   
145      wagon          4wd           front        96.9  0.834214  ...   

     compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
9                  8.8       101.0    5800.0       23          29  16430.0   
68                 8.3       155.0    4750.0       16          18  34184.0   
145                7.7       111.0    4800.0       23          23  11694.0   

    city-L/100km  horsepower-binned  diesel  gas  
9      10.217391                Low       0    1  
68     14.687500             Medium       0    1  
145    10.217391             Medium       0    1  

[3 rows x 29 columns]

## 11. monitor the amount of data used in each of the columns of your data frame using .memory_usage parameter
To limit the runtimes of your programs you must be vigilant over the size of the objects you are creating in your program, and also the number of objects you are creating. It returns as series the total number of bytes of memory that the data frame is currently occupying in my memory right now.

In [17]:
df.memory_usage()

Index                 128
symboling            1608
normalized-losses    1608
make                 1608
aspiration           1608
num-of-doors         1608
body-style           1608
drive-wheels         1608
engine-location      1608
wheel-base           1608
length               1608
width                1608
height               1608
curb-weight          1608
engine-type          1608
num-of-cylinders     1608
engine-size          1608
fuel-system          1608
bore                 1608
stroke               1608
compression-ratio    1608
horsepower           1608
peak-rpm             1608
city-mpg             1608
highway-mpg          1608
price                1608
city-L/100km         1608
horsepower-binned    1608
diesel               1608
gas                  1608
dtype: int64

#### Hope this blog serves in making your work easier as a Data Scientist 